flask_test_app/app.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Flask application made for integration tests."""
+
+import time
+from urllib.parse import urlparse
+
+import psycopg
+import pymysql
+import pymysql.cursors
+from flask import Flask, g, jsonify, request
+
+app = Flask(__name__)
+app.config.from_prefixed_env()
+
+
+def get_mysql_database():
+    """Get the mysql db connection."""
+    if "mysql_db" not in g:
+        if app.config.get("MYSQL_DB_CONNECT_STRING"):
+            uri_parts = urlparse(app.config["MYSQL_DB_CONNECT_STRING"])
+            g.mysql_db = pymysql.connect(
+                host=uri_parts.hostname,
+                user=uri_parts.username,
+                password=uri_parts.password,
+                database=uri_parts.path[1:],
+                port=uri_parts.port,
+            )
+        else:
+            return None
+    return g.mysql_db
+
+
+def get_postgresql_database():
+    """Get the postgresql db connection."""
+    if "postgresql_db" not in g:
+        if app.config.get("POSTGRESQL_DB_CONNECT_STRING"):
+            uri_parts = urlparse(app.config.get("POSTGRESQL_DB_CONNECT_STRING"))
+            g.postgresql_db = psycopg.connect(
+                host=uri_parts.hostname,
+                user=uri_parts.username,
+                password=uri_parts.password,
+                dbname=uri_parts.path[1:],
+                port=uri_parts.port,
+            )
+        else:
+            return None
+    return g.postgresql_db
+
+
+@app.teardown_appcontext
+def teardown_database(_):
```

Comment:
Aren't there libraries that do all of this for you? Like https://flask-mysql.readthedocs.io/en/stable/

Reply:
For our usage, I could even remove this function. I like having things explicit when they are so simple and in tests.

Reply:
it isn't just this it is also the rest of the other database functions, most users would also use flask-mysql so a good side effect is to check what the experience would be with that

Reply:
I think most would use sqlalchemy (7.2k stars) or pymysql (7.3k stars) instead of flask-mysql (116 stars). I choose to not use sqlalchemy to keep things simple

Reply:
There is also https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/ which is the library that I should have linked to initially

Reply:
I'd go with simplicity in that case - it's clear and does what it does. Using some ORM and then in tests might be too much, it would also hint at our preference (for instance sqlalchemy), where a developer should chose whatever they want, if they want to use an ORM in the first place.

Reply:
We can leave it like this for now